### ISSUE TITLE
Playing a song using MediaSessionService randomly turns on and off bluetooth for some reason on Android 11.

### ISSUE THREAD

**Issue by debz-eight** [androidx/media#1552]
I have used MediaSessionService, MediaSession.Callback() and MediaController according to docs. The setup is working fine in all Android versions except Android 11.

I have tested on these devices:
Pixel 5 
Xiaomi Redmi Note 8 Pro.

Error Log:

```
2024-07-18 18:18:46.249 19916-19944 BluetoothAdapter        com.network.eight.android            D  onBluetoothServiceUp: android.bluetooth.IBluetooth$Stub$Proxy@76be954
2024-07-18 18:18:47.596 22671-22671 MediaBrowser            com.android.bluetooth                E  Failed binding to service ComponentInfo{com.network.eight.android/com.network.eight.services.SongPlayerService}
2024-07-18 18:18:47.741 19916-22252 CCodecBuffers           com.network.eight.android            D  [c2.android.aac.decoder#79:1D-Input.Impl[N]] codec released a buffer owned by client (index 0)
2024-07-18 18:18:47.746 19916-22252 CCodecBuffers           com.network.eight.android            D  [c2.android.aac.decoder#79:1D-Input.Impl[N]] codec released a buffer owned by client (index 2)
2024-07-18 18:18:49.142 19916-22248 BufferPoolAccessor2.0   com.network.eight.android            D  bufferpool2 0xb400007a21a94b68 : 4(32768 size) total buffers - 4(32768 size) used buffers - 1/6 (recycle/alloc) - 5/1554 (fetch/transfer)
2024-07-18 18:18:57.616 19916-19944 BluetoothAdapter        com.network.eight.android            D  onBluetoothServiceDown: android.bluetooth.IBluetooth$Stub$Proxy@76be954
2024-07-18 18:18:57.623 19916-19944 BluetoothAdapter        com.network.eight.android            D  onBluetoothServiceDown: null
2024-07-18 18:18:58.175 19916-19944 BluetoothAdapter        com.network.eight.android            D  onBluetoothServiceUp: android.bluetooth.IBluetooth$Stub$Proxy@e9bd5ef
2024-07-18 18:18:59.534 22804-22804 MediaBrowser            com.android.bluetooth                E  Failed binding to service ComponentInfo{com.network.eight.android/com.network.eight.services.SongPlayerService}
2024-07-18 18:19:09.554 19916-19944 BluetoothAdapter        com.network.eight.android            D  onBluetoothServiceDown: android.bluetooth.IBluetooth$Stub$Proxy@e9bd5ef
2024-07-18 18:19:09.560 19916-19944 BluetoothAdapter        com.network.eight.android            D  onBluetoothServiceDown: null
2024-07-18 18:19:10.150 19916-19944 BluetoothAdapter        com.network.eight.android            D  onBluetoothServiceUp: android.bluetooth.IBluetooth$Stub$Proxy@ecade29
2024-07-18 18:19:11.582 22913-22913 MediaBrowser            com.android.bluetooth                E  Failed binding to service ComponentInfo{com.network.eight.android/com.network.eight.services.SongPlayerService}
```

Any idea what this might be issue.

**Comment by debz-eight** [comment]
@marcbaechinger Any idea with why this is occurring. Can you provide with any lead on this please?

**Comment by debz-eight** [comment]
Tested the same on Pixel 3 Android 11. 
Screenrecording: [Drive](https://drive.google.com/file/d/1_vbEGcXCDCJWDAxBkY2swdY7OIZQJBHc/view?usp=sharing)

I am streaming HLS media that ends with m3u8.

```
 val mediaItems = getSongList().map { song ->
                        MediaItem.Builder().setUri(song.audio).setMediaId(song.audio).setMediaMetadata(
                            MediaMetadata.Builder().setFolderType(MediaMetadata.FOLDER_TYPE_ALBUMS)
                                .setArtworkUri(Uri.parse(song.bannerSquare?.md ?: song.banner2))
                                .setArtist(getSeriesName()).setTitle(song.name)
                                .setExtras(putMediaItemsData()).build()
                        ).build()
                    }
                    controller.setMediaItems(mediaItems)
                    controller.seekTo(
                        getCurrentSongPosition(),
                        getCurrentPlayingSongProgressDataInMs()
                    )
```

This is being received by onAddMediaItems() and the player is preparing and playing it.

Here is my manifest declaration:

```
<service
            android:name=".services.SongPlayerService"
            android:exported="false"
            android:foregroundServiceType="mediaPlayback">
            <intent-filter>
                <action android:name="androidx.media3.session.MediaSessionService" />
                <action android:name="android.media.browse.MediaBrowserService" />
            </intent-filter>
        </service>
```

**Comment by debz-eight** [comment]
One thing that I can't comprehend is how is it affecting bluetooth when there is no permission for bluetooth in my app.

**Comment by debz-eight** [comment]
After some more debugging, I found that, when setting a single media item, the issue does not persist. When I am setting a List<MediaItems> via controller using controller.setMediaItems() then only this issue is occuring

**Comment by debz-eight** [comment]
@marcbaechinger Can you please have a look at this issue? Right now this is a huge blocker for our company to release the media3 along with downloads features. Our app: [Eight - Audio Done Better](https://play.google.com/store/apps/details?id=com.network.eight.android) is a heavily audio based app and we are blocked on this for the issue on Android 11. We can't make a release for the same. Would be really helpful if you can take a look at it.

**Comment by marcbaechinger** [comment]
Thanks for your report. Difficult to say what's wrong I'm afraid. Media3 isn't doing something Bluetooth specific. The screencast alludes BT is crashing on the OS level as it seems to disconnect.

Can you repro the same with the session demo app? Is this happening on a different device in the same way? I tried the session demo app with a Pixel 5 with Android 11 flashed on it and I couldn't see something going wrong. 

If you can repro the same with the demo app, then please do the bug report for the demo app failure right after it happened and upload here or send it to android-media-github@google.com using a subject in the format "Issue #1552". Please also update this issue to indicate you’ve done this.

The error log above tells that, Bluetooth can't bind to your service. This should make BT crash though.

When you have expose your service as ` MediaBrowserService` Bluetooth attempts to connect with a platform `MediaController` and `MediaBrowser` against your service. You can see this for instance when your app is installed and you restart the device. When restarting a Pixel 5 running Android 11,  I see the following logs statements that I added in the methods of `MediaLibrarySession.Callback` (`onConnect`, `onGetLibraryRoot` and `onGetChildren`).

The log show BT binding against the service and calling domain methods.

```
---------------------------- PROCESS STARTED (4179) for package androidx.media3.demo.session ----------------------------
2024-07-23 18:59:11.025  4179-4179  bt-connect              androidx.media3.demo.session         D  onBind to android.media.browse.MediaBrowserService
2024-07-23 18:59:11.156  4179-4179  app-callback            androidx.media3.demo.session         D  onConnect called for com.android.bluetooth
2024-07-23 18:59:11.157  4179-4179  app-callback            androidx.media3.demo.session         D  onGetLibraryRoot called for com.android.bluetooth
2024-07-23 18:59:11.276  4179-4179  app-callback            androidx.media3.demo.session         D  onConnect called for com.android.bluetooth
2024-07-23 18:59:11.277  4179-4179  app-callback            androidx.media3.demo.session         D  onGetLibraryRoot called for com.android.bluetooth
2024-07-23 18:59:11.287  4179-4179  app-callback            androidx.media3.demo.session         D  onGetChildren called for com.android.bluetooth with parentId=[rootID]
---------------------------- PROCESS ENDED (4179) for package androidx.media3.demo.session ----------------------------
```

However, I don't think that rejecting these calls would make BT crash because this is common and the BT code anticipates this. For instance the search app rejects connection and BT doesn't crash:

```
2024-07-23 19:07:39.436  3651-3651  MediaBrowserService     com....android.googlequicksearchbox  I  No root for client com.android.bluetooth from service android.service.media.MediaBrowserService$ServiceBinder$1
2024-07-23 19:07:39.439  2140-2140  MediaBrowser            com.android.bluetooth                E  onConnectFailed for ComponentInfo{com.google.android.googlequicksearchbox/com.google.android.libraries.assistant.auto.tng.media.mediabrowser.NewsMediaBrowserService}
2024-07-23 19:07:39.445  2140-2140  AvrcpBrows...yerWrapper com.android.bluetooth                W  onConnectionFailed: Connection Failed with com.google.android.googlequicksearchbox
```

Going forward I'd suggest:

0. To rule out: Make sure to accept the connection from BT in `onConnect` and `onGetLirbaryRoot`, these two are essential for letting a legacy/platform `MediaBrowser` connect that BT is using.
1. Try to repro with the session demo app on different devices on Android 11
2. Send us a bug report as explained above for a repro.

**Comment by debz-eight** [comment]
Really thanks for the reply. I will try 0th suggestion you said and report back.

**Comment by debz-eight** [comment]
@marcbaechinger I found the issue, its pretty weird. So I was setting extras in MediaMetadata and inside one of those extras' fields, I had an Array List of a data class which I was passing using putParcelableArrayList. The data class had @Parcelize and Parcelable.

The issue was com.android.bluetooth could not find the data class while unmarshalling.

To fix it, I simply use gson.toJson and converted the list to json and then retrieved it via gson.fromJson and it fixed the issue. Also I did not have to put any code to accept bluetooth connection.

**Comment by debz-eight** [comment]
@marcbaechinger I found the issue, its pretty weird. So I was setting extras in MediaMetadata and inside one of those extras' fields, I had an Array List of a data class which I was passing using putParcelableArrayList. The data class had @Parcelize and Parcelable.

The issue was com.android.bluetooth could not find the data class while unmarshalling.

To fix it, I simply use gson.toJson and converted the list to json and then retrieved it via gson.fromJson and it fixed the issue. Also I did not have to put any code to accept bluetooth connection.
